### PR TITLE
Fuse MLA DOWN projection GEMMs

### PR DESCRIPTION
**Pull request by cjld**
dev PR: #2960 

# What does this PR do ?
Add cli args `--mla-down-proj-fusion` and `mla_down_proj_fusion` to the transformer config for the fusion.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by maanug-nv** [comment]
is there a PR to dev that we can link? i don't see the same changes on dev branch currently?

File: megatron/core/models/gpt/gpt_layer_specs.py [hunk 1]
```diff
@@ -235,6 +242,31 @@ def get_gpt_layer_with_transformer_engine_spec(
 
     if multi_latent_attention:
         assert qk_l2_norm is False, "qk_l2_norm is not supported with MLA."
+
+        fuse_input_layernorm = (
```
Comment (maanug-nv) [inline megatron/core/models/gpt/gpt_layer_specs.py]: can `fuse_input_layernorm` just be an argument to this function? that seems a more consistent UX than adding environment variables for any adjustment to the spec.

File: megatron/core/models/gpt/gpt_layer_specs.py [hunk 2]
```diff
@@ -171,6 +171,13 @@ def get_gpt_layer_with_inference_spec(
         )
 
 
+import os
+# NOTE: These toggles are intentionally environment-driven for experimentation.
+# They are expected to be replaced by explicit config options.
```
Comment (jaredcasper) [inline megatron/core/models/gpt/gpt_layer_specs.py]: Please replace them now. We do not use environment variables to configure anything in mcore.

**Comment by maanug-nv** [comment]
Hi @cjld , are you able to update this PR to remove usage of env vars? I think that's the main blocker

File: megatron/core/models/gpt/gpt_layer_specs.py (hunk 1 above)
Comment (cjld) [reply megatron/core/models/gpt/gpt_layer_specs.py]: ok, added fuse_mla_input_layernorm

File: megatron/core/models/gpt/gpt_layer_specs.py (hunk 2 above)
Comment (cjld) [reply megatron/core/models/gpt/gpt_layer_specs.py]: ok, replaced -> --qkv-down-proj-fusion

**Comment by cjld** [comment]
Hi, @maanug-nv , I updated the PR. If the PR still has problems, please tell me.

**Comment by cjld** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3039/commits/e245e191724de4766e8e4b4f8af742a5810a7fa8

File: megatron/core/transformer/multi_latent_attention.py [hunk 3]
```diff
@@ -411,29 +424,31 @@ def __init__(
...
-                tp_comm_buffer_name='q_down_proj',
-                skip_weight_param_allocation=False,
-                tp_group=(
-                    pg_collection.tp
-                    if q_down_proj_kwargs.get('parallel_mode') != 'duplicated'
-                    else None
-                ),
-                **q_down_proj_kwargs,
-            )
+            if not self.config.qkv_down_proj_fusion:
```
Comment (jaredcasper) [inline megatron/core/transformer/multi_latent_attention.py]: Shouldn't this be handled by the spec?

File: megatron/core/transformer/multi_latent_attention.py [hunk 4]
```diff
@@ -456,29 +471,55 @@ def __init__(
...
-            tp_comm_buffer_name='kv_down_proj',
-            skip_weight_param_allocation=False,
-            tp_group=(
-                pg_collection.tp
-                if kv_down_proj_kwargs.get('parallel_mode') != 'duplicated'
-                else None
-            ),
-            **kv_down_proj_kwargs,
-        )
+        if not self.config.qkv_down_proj_fusion:
```
Comment (jaredcasper) [inline megatron/core/transformer/multi_latent_attention.py]: likewise here.

File: megatron/core/transformer/multi_latent_attention.py (hunk 3 above)
Comment (cjld) [reply megatron/core/transformer/multi_latent_attention.py]: fusion now control by spec, see [1d97497](https://github.com/NVIDIA/Megatron-LM/pull/3039/commits/1d97497ca0eafe7c0a06b18ee4621baedd1a061d)

File: megatron/core/transformer/multi_latent_attention.py (hunk 4 above)
Comment (cjld) [reply megatron/core/transformer/multi_latent_attention.py]: same as above

File: megatron/core/transformer/multi_latent_attention.py [hunk 5]
```diff
@@ -460,29 +477,55 @@ def __init__(
...
-            tp_comm_buffer_name='kv_down_proj',
-            skip_weight_param_allocation=False,
-            tp_group=(
-                pg_collection.tp
-                if kv_down_proj_kwargs.get('parallel_mode') != 'duplicated'
-                else None
-            ),
-            **kv_down_proj_kwargs,
-        )
+        if submodules.linear_qkv_down_proj is None:
```
Comment (jaredcasper) [inline megatron/core/transformer/multi_latent_attention.py]: Can we get rid of this if/else. We really want to avoid things like this in the modules. Either using the spec and IdentityOps, or abstracting things out into different classes derived from a common base (see for example SelfAttention vs CrossAttention). Maybe FusedMLASelfAttention or something?

Comment (cjld) [reply megatron/core/transformer/multi_latent_attention.py]: Okay, I'll finish the refactoring after the Chinese new year

Comment (cjld) [reply megatron/core/transformer/multi_latent_attention.py]: Hi, jaredcasper. 
Thanks for the review — I refactored the MLA fusion path to follow the same pattern as separate attention implementations (e.g., class-level specialization selected in the spec), instead of adding runtime branching in the core attention module.

What changed
* Added a dedicated `FusedMLASelfAttention` class and kept MLASelfAttention as the unfused baseline.
* Switched selection to spec-time via ModuleSpec in GPT TE layer spec (mla_fusion decides which class is instantiated).
* Kept fusion as a single best-effort flag (`--mla-fusion`): it enables fused q/kv down-projection and fused input layernorm when backend support exists, otherwise falls back gracefully.
* Added checkpoint compatibility logic so fused implementation can interoperate with legacy unfused checkpoint formats.

This should address the concern about runtime conditional complexity in the core module while preserving the intended fusion behavior.

File: megatron/core/models/gpt/gpt_layer_specs.py [hunk 6]
```diff
@@ -248,6 +253,47 @@ def get_gpt_layer_with_transformer_engine_submodules(
             if qk_layernorm
             else backend.column_parallel_linear()
         )
+
+        if mla_down_proj_fusion:
+            fuse_input_layernorm = (
+                mla_down_proj_fusion and backend.column_parallel_layer_norm_linear() is not None
```
Comment (ericharper) [inline megatron/core/models/gpt/gpt_layer_specs.py]: do we need mla_down_proj_fusion here? we're already in the conditional?

**Comment by ericharper** [comment]
Can we add unit tests for this feature?

**Comment by cjld** [comment]
> Can we add unit tests for this feature?

OK

**Comment by cjld** [comment]
Unittests added.

File: gpt_builders.py [hunk 7]
```diff
@@ -137,6 +137,7 @@ def _get_transformer_layer_spec(use_te, config):
             use_te_activation_func=config.use_te_activation_func,
             use_kitchen_attention=config.use_kitchen_attention,
             kitchen_attention_backend=config.kitchen_attention_backend,
+            mla_down_proj_fusion=args.mla_down_proj_fusion,
```
Comment (gdengk) [inline gpt_builders.py]: Can we make sure this option could be set through transformer_config as well?
mla_down_proj_fusion: bool = False in transformer_config.py is a dead field.

Comment (yaox12) [reply gpt_builders.py]: Taking a closer look at the code, we should replace all these `args.` with `config.` right? That's to say, all these `num_experts`/`moe_grouped_gemm`/`qk_layernorm` ... options are only set through args but not `transformer_config.py`? I'm a bit shocked by this.

Comment (yaox12) [reply gpt_builders.py]: Also, we need another PR to the Megatron-Bridge to correctly parse this option.

Comment (gdengk) [reply gpt_builders.py]: Megatron Bridge calls the `get_gpt_decoder_block_spec`  (from megatron.core.models.gpt.gpt_layer_specs import get_gpt_decoder_block_spec). So actually it have no access to the function in gpt_builder.py. The func in gpt_builders.py is limited to Megatron-lm usage only so I think options starting with args. should still be fine.
Just talked with Dun, I think we need to pass the option correct in the get_gpt_decoder_block_spec from megatron/core/models/gpt/gpt_layer_specs.py .

**Comment by cjld** [comment]
> Can we add unit tests for this feature?

Hi, @ericharper , could you please review this unittest?

Merged FusedMLA unit tests into `test_multi_latent_attention.py`, covering:

- **Constructor & weight shape**: Verifies `FusedMLASelfAttention` instantiation and fused down-projection weight dimensions (`q_lora_rank + kv_lora_rank + qk_pos_emb_head_dim`).
- **Forward pass**: Tests GPU forward in both fp32 and bf16, parametrized over `rope_type` (`yarn`/`rope`).
- **QKV split**: Validates `_qkv_down_projection` output shapes for q and kv branches.
- **Backward pass**: Confirms gradient flow through `linear_qkv_down_proj` and input.
- **State dict compatibility**: Loads unfused state dict into fused model and verifies weight fusion; checks `sharded_state_dict()` splits fused weight back into separate q/kv keys.
- **Error handling**: Asserts `q_lora_rank=None` raises `AssertionError`.

All 14 tests pass.
